### PR TITLE
Added template and scripts for generating rustdocs

### PR DESCRIPTION
**Pull request by jimmychu0807**
This PR adds capability of:
- Generating a rustdocs index template page based on a template (/.maintain/docs-index-tpl.ejs). [An example here](https://jimmychu0807.github.io/advent-of-code-2020/).
- Generating the rustdocs that fit a certain branch name defined in `$RUSTDOCS_DEPLOY_REFS` in `.gitlab-ci.yml`. 

Note that in order to generate all the rustdocs version specified in `$RUSTDOCS_DEPLOY_REFS`, those branches/tags need to be pushed into the repo at least once with this CI script present.

I will write a CI task that can generate all rustdocs version in a separate PR. This CI task is to be triggered manually on Gitlab side.

@TriplEight Please review the CI script.

This is part of the task for https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/issues/1091
cc @imadarai @sacha-l

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
   # `needs:` can be removed after CI image gets nonroot. In this case `needs:` stops other
   # artifacts from being dowloaded by this job.
   needs:
     - job:                         build-rustdoc
       artifacts:                   true
-  script:
-    - rm -rf /tmp/*
+  script: |
+    rm -rf /tmp/*
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:760]:
```suggestion
  script:
    - rm -rf /tmp/*
```

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
+    git config user.email "devops-team@parity.io"
+    git config user.name "${GITHUB_USER}"
+    git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
+    git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
+    git fetch origin gh-pages
+
+    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
+    yarn global add ejs
+    ejs ${INDEX_TPL} -i "{\"deploy_refs\": \"${RUSTDOCS_DEPLOY_REFS}\", \"repo_name\": \"${CI_PROJECT_NAME}\", \"latest\": \"${LATEST}\"}" > /tmp/index.html
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:772]:
```suggestion
    - rm -rf .git/config
    - git config user.email "devops-team@parity.io"
    - git config user.name "${GITHUB_USER}"
    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
    - git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
    - git fetch origin gh-pages
    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
    - yarn global add ejs
    - ejs ${INDEX_TPL} -i "{\"deploy_refs\": \"${RUSTDOCS_DEPLOY_REFS}\", \"repo_name\": \"${CI_PROJECT_NAME}\", \"latest\": \"${LATEST}\"}" > /tmp/index.html
```

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
-    - git fetch origin gh-pages
+    rm -rf .git/config
+    git config user.email "devops-team@parity.io"
+    git config user.name "${GITHUB_USER}"
+    git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
+    git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
+    git fetch origin gh-pages
+
+    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
+    yarn global add ejs
```
Comment (TriplEight) [inline .gitlab-ci.yml:770]: is there any way to not install something on every CI run?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
+    # Also ensure $RUSTDOCS_DEPLOY_REFS is non-space
+    if [[ ! -z ${RUSTDOCS_DEPLOY_REFS// } ]]; then
+      for FILE in *; do
+        if [[ ! " $RUSTDOCS_DEPLOY_REFS " =~ " $FILE " ]]; then
+          echo "Removing ${FILE}..."
+          rm -rf $FILE
+        fi
+      done
+    fi
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:788]:
```suggestion
    - cp -r ./crate-docs/ /tmp/doc/
    - cp README.md /tmp/doc/
    - git checkout gh-pages
    # Remove directories no longer necessary, as specified in $RUSTDOCS_DEPLOY_REFS.
    # Also ensure $RUSTDOCS_DEPLOY_REFS is non-space
    - if [[ ! -z ${RUSTDOCS_DEPLOY_REFS// } ]]; then
        for FILE in *; do
          if [[ ! " $RUSTDOCS_DEPLOY_REFS " =~ " $FILE " ]]; then
            echo "Removing ${FILE}..."
            rm -rf $FILE
          fi
        done
      fi
```

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
+    # Move the index page & built back
+    mv -f /tmp/index.html .
+    # Ensure the destination dir doesn't exist.
+    rm -rf ${CI_COMMIT_REF_NAME}
+    mv -f /tmp/doc ${CI_COMMIT_REF_NAME}
+    # Add the symlink
+    [[ -e "$LATEST" ]] && ln -sf "${LATEST}" latest
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:796]:
```suggestion
    # Move the index page & built back
    - mv -f /tmp/index.html .
    # Ensure the destination dir doesn't exist.
    - rm -rf ${CI_COMMIT_REF_NAME}
    - mv -f /tmp/doc ${CI_COMMIT_REF_NAME}
    # Add the symlink
    - [[ -e "$LATEST" ]] && ln -sf "${LATEST}" latest

```

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
     # Upload files
-    - git add --all --force
+    git add --all --force
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:799]:
```suggestion
    - git add --all --force
```

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
     # `git commit` has an exit code of > 0 if there is nothing to commit.
     # This causes GitLab to exit immediately and marks this job failed.
     # We don't want to mark the entire job failed if there's nothing to
     # publish though, hence the `|| true`.
-    - git commit -m "Updated docs for ${CI_COMMIT_REF_NAME}" ||
-        echo "___Nothing to commit___"
-    - git push origin gh-pages --force
+    git commit -m "Updated docs for ${CI_COMMIT_REF_NAME}" || echo "___Nothing to commit___"
+    git push origin gh-pages --force
+
```
Comment (TriplEight) [inline .gitlab-ci.yml:806]:
```suggestion
    - git commit -m "___Updated docs for ${CI_COMMIT_REF_NAME}___" ||
        echo "___Nothing to commit___"
    - git push origin gh-pages --force
```

**Review by TriplEight — CHANGES_REQUESTED** [review]
Please return the lines that you didn't change.
Otherwise, it could've been a bash script with stages commented with `echo`.

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
   stage:                           publish
   <<:                              *kubernetes-env
   <<:                              *vault-secrets
-  image:                           paritytech/tools:latest
+  image:                           node:16
   variables:
     GIT_DEPTH:                     100
+    # --- Following variables are for rustdocs deployment ---
+    # Space separated values of branches and tags to generate rustdocs
+    RUSTDOCS_DEPLOY_REFS:   "master monthly-2021-09+1 monthly-2021-08 v3.0.0"
```
[the remark this answers is not in the document]
Comment (jimmychu0807) [reply .gitlab-ci.yml:737]: These are the tags / branches from github substrate repo that we want to generate a version of rustdocs. So the above means we want to generate docs for:

- [master](https://github.com/paritytech/substrate/tree/master)
- [monthly-2021-09+1](https://github.com/paritytech/substrate/tree/monthly-2021-09+1)
- [monthly-2021-08](https://github.com/paritytech/substrate/tree/monthly-2021-08)
- [v3.0.0](https://github.com/paritytech/substrate/tree/v3.0.0)

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -728,44 +729,81 @@ publish-rustdoc:
...
   rules:
     - if: $CI_PIPELINE_SOURCE == "pipeline"
       when: never
-    - if: $CI_PIPELINE_SOURCE == "web" && $CI_COMMIT_REF_NAME == "master"
+    # Manually add branches based on $RUSTDOCS_DEPLOY_REFS
+    # Maybe there is a smarter way?
     - if: $CI_COMMIT_REF_NAME == "master"
+    - if: $CI_COMMIT_REF_NAME == "monthly-2021-09+1"
+    - if: $CI_COMMIT_REF_NAME == "monthly-2021-08"
+    - if: $CI_COMMIT_REF_NAME == "v3.0.0"
```
[the remark this answers is not in the document]
Comment (jimmychu0807) [reply .gitlab-ci.yml:751]: Actually, from docs team perspective, we want to keep a few version of rustdocs, but not necessarily all version docs. So I want a finer control of which version of docs we want to publish and use `$RUSTDOCS_DEPLOY_REFS` for reference. We also want to have 2-3 of the "monthly" Substrate docs out.

File: .gitlab-ci.yml [hunk 10]
```diff
@@ -747,23 +759,41 @@ publish-rustdoc:
     - rm -rf .git/config
     - git config user.email "devops-team@parity.io"
     - git config user.name "${GITHUB_USER}"
-    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/substrate.git"
+    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
     - git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
     - git fetch origin gh-pages
+    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
+    - yarn global add ejs
```
Comment (jimmychu0807) [inline .gitlab-ci.yml:766]: probably we can do caching here, but not sure where to cache.

File: .gitlab-ci.yml [hunk 11]
```diff
@@ -747,23 +759,42 @@ publish-rustdoc:
...
+    - |
```
Comment (jimmychu0807) [inline .gitlab-ci.yml:767]: If we combine 767 - 768, yaml doesn't parse the ':' and '"' correctly, or we need some tricky character escaping. So just making this multiline, and parse everything as string.

File: .gitlab-ci.yml (hunk 3 above)
Comment (jimmychu0807) [reply .gitlab-ci.yml]: Probably we can do caching here, but not sure where to cache.

File: .gitlab-ci.yml (hunk 10 above)
Comment (jimmychu0807) [reply .gitlab-ci.yml:767]: dup'ed.

**Comment by jimmychu0807** [comment]
Most items are resolved. This https://github.com/paritytech/substrate/pull/9785#discussion_r709857107 may need some discussions with you.

File: .gitlab-ci.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (TriplEight) [reply .gitlab-ci.yml:751]: I can't see why automating is a problem here. On the contrary, for every new release, you will have to create a new PR adding a new `monthly*`, why not just
`- if: $CI_COMMIT_REF_NAME =~ /^monthly-(202[1-9])-(1[0-2]|0[1-9]|\d).*$/`

File: .gitlab-ci.yml (hunk 11 above)
Comment (TriplEight) [reply .gitlab-ci.yml:768]: for this you could put the command into single quotes: 
`- 'command'`

File: .gitlab-ci.yml [hunk 12]
```diff
@@ -747,23 +760,42 @@ publish-rustdoc:
...
+        done
+      fi
+    # Move the index page & built back
+    - mv -f /tmp/index.html .
+    # Ensure the destination dir doesn't exist.
+    - rm -rf ${CI_COMMIT_REF_NAME}
+    - mv -f /tmp/doc ${CI_COMMIT_REF_NAME}
+    # Add the symlink
+    - |
+      [[ -e "$LATEST" ]] && ln -sf "${LATEST}" latest
```
Comment (TriplEight) [inline .gitlab-ci.yml:791]:
```suggestion
    - [[ -e "$LATEST" ]] && ln -sf "${LATEST}" latest
```

**Review by TriplEight — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml (hunk 12 above)
Comment (jimmychu0807) [reply .gitlab-ci.yml:791]: The problem of '- command' is that it won't interpolate the variable `${RUSTDOCS_DEPLOY_REFS}`. These values will become just plain string.

**Comment by jimmychu0807** [comment]
@TriplEight The latest commit should resolve all issues. 
Note the `ejs` line is a bit ugly if not using `|` at front. `'` is not possible as it won't interpolate `$VAR`.

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -728,14 +729,23 @@ publish-rustdoc:
...
+    INDEX_TPL:                     ".maintain/docs-index-tpl.ejs"
+    # Where the `/latest` symbolic link links to. One of the $RUSTDOCS_DEPLOY_REFS value.
+    LATEST:                        "monthly-2021-09+1"
   rules:
     - if: $CI_PIPELINE_SOURCE == "pipeline"
       when: never
     - if: $CI_PIPELINE_SOURCE == "web" && $CI_COMMIT_REF_NAME == "master"
     - if: $CI_COMMIT_REF_NAME == "master"
+    - if: $CI_COMMIT_REF_NAME =~ /^monthly-20[0-9]{2}-[0-9]{2}.*$/  # to support: monthly-2021-09+1
+    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/
```
Comment (TriplEight) [inline .gitlab-ci.yml:748]:
```suggestion
    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/              # i.e. v1.0, v2.1rc1
```
let's comment this too

File: .gitlab-ci.yml [hunk 14]
```diff
@@ -747,23 +757,40 @@ publish-rustdoc:
     - rm -rf .git/config
     - git config user.email "devops-team@parity.io"
     - git config user.name "${GITHUB_USER}"
-    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/substrate.git"
+    - git config remote.origin.url "https://${GITHUB_TOKEN}@github.com/paritytech/${CI_PROJECT_NAME}.git"
     - git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*"
     - git fetch origin gh-pages
+    # Install `ejs` and generate index.html based on RUSTDOCS_DEPLOY_REFS
+    - yarn global add ejs
+    - "ejs ${INDEX_TPL} -i \"{\\\"deploy_refs\\\":\\\"${RUSTDOCS_DEPLOY_REFS}\\\",\\\"repo_name\\\":\\\"${CI_PROJECT_NAME}\\\",\\\"latest\\\":\\\"${LATEST}\\\"}\" > /tmp/index.html"
```
Comment (TriplEight) [inline .gitlab-ci.yml:765]: This won't work surrounded with `" "` as it has a lot of the same brackets inside, please use `' ' `

File: .gitlab-ci.yml [hunk 15]
```diff
@@ -747,23 +757,40 @@ publish-rustdoc:
...
+          fi
+        done
+      fi
+    # Move the index page & built back
+    - mv -f /tmp/index.html .
+    # Ensure the destination dir doesn't exist.
+    - rm -rf ${CI_COMMIT_REF_NAME}
+    - mv -f /tmp/doc ${CI_COMMIT_REF_NAME}
+    # Add the symlink
+    - "[[ -e \"$LATEST\" ]] && ln -sf \"${LATEST}\" latest"
```
Comment (TriplEight) [inline .gitlab-ci.yml:786]: This won't work surrounded with `" "` as it has a lot of the same brackets inside, please use `' ' `

**Review by TriplEight — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml [hunk 16]
```diff
@@ -728,42 +729,73 @@ publish-rustdoc:
...
   # artifacts from being dowloaded by this job.
   needs:
     - job:                         build-rustdoc
       artifacts:                   true
   script:
+    # If $CI_COMMIT_REF_NAME doesn't match one of $RUSTDOCS_DEPLOY_REFS space-separated values, we
+    # exit immediately.
+    # Putting spaces at the front and back to ensure we are not matching just any substring, but the
+    # whole space-separated value.
+    - '[[ " ${RUSTDOCS_DEPLOY_REFS} " =~ " ${CI_COMMIT_REF_NAME} " ]] || exit 0'
```
Comment (jimmychu0807) [inline .gitlab-ci.yml:759]: @TriplEight Added this line to check if it matches and exit early if not.

**Review by TriplEight — APPROVED** [review]

**Comment by jimmychu0807** [comment]
@rcny Hi Vladimir, will be great if you could help review this gitlab CI script, and if everything look okay to you, approve it. Thanks!

**Comment by rcny** [comment]
LGTM

**Review by rcny — APPROVED** [review]

**Pull request merged**